independent-projects/arc/runtime/src/main/java/io/quarkus/arc/InactiveBeanException.java

```diff
@@ -0,0 +1,12 @@
+package io.quarkus.arc;
+
+import jakarta.enterprise.inject.InjectionException;
+
+/**
+ * Indicates that an inactive bean was injected (or dynamically looked up).
+ */
+public class InactiveBeanException extends InjectionException {
```

Comment:
So first, having this exception is great, and allows me to do everything I need.

That being said, I would like to point out two things:

* Right now the only way to check if a bean is inactive is [to fetch it](https://github.com/quarkusio/quarkus/pull/41929/files#diff-93d12c860e8b94edea17693d4d0ac031cef765d2c74cc758f283839e4ae00af9R33-R38). Do we want to introduce `InjectableInstance#isActive()` to allow easy programmatic lookup?
* Right now the only way to retrieve a bean optionally, only if it's active, is to [catch the exception and return null/optional](https://github.com/quarkusio/quarkus/pull/41929/files#diff-93d12c860e8b94edea17693d4d0ac031cef765d2c74cc758f283839e4ae00af9R25-R31). And when retrieving multiple beans (e.g. all instances of `AgroalDataSource`) that can get rather impractical; one would most likely have to use `InjectableInstance#handles`. Do we want to add an `Active` qualifier that could be passed to `BeanContainer#select()` to tell Arc to ignore inactive beans?

Note I'm not suggesting to to anything more in this PR, it can definitely be a follow-up issue -- which may even die a slow death :)

Reply:
I don't think an `@Active` qualifier is a good idea, because all other injection points would suddenly also have to use it.

Previously, I had a method `InjectableBean.checkActive()`, which I deleted but could add again. Perhaps in a slightly altered form: `boolean isActive()`. An `InjectableBean` is available from `InstanceHandle`. That seems rather straightforward to me:

```java
@Inject
@Any
Instance<AgroalDataSource> ds;

...

for (InstanceHandle handle : ds.handles()) {
    if (handle.getBean().isActive()) {
        AgroalDataSource datasource = handle.get();
        ...
    }
}
```

Doesn't look all that bad to me?

Reply:
> Previously, I had a method `InjectableBean.checkActive()`, which I deleted but could add again. Perhaps in a slightly altered form: `boolean isActive()`.

LGTM

> I don't think an @Active qualifier is a good idea, because all other injection points would suddenly also have to use it.

I suppose you're saying that because the bean would (per spec) lose the `@Default` qualifier?

Since `@Active` is not in the spec, I'd say you're free to declare that it has special behavior, e.g. it can't be applied to a bean explicitly, and it doesn't lead to the `@Default` qualifier being dropped.

After all, there's precedent with `@Any`: that qualifier is added automatically on every bean, and doesn't affect `@Default`.

> Doesn't look all that bad to me?

It's less practical than a one-liner, but I agree it's reasonable. Probably the best compromise we can get if the `@Active` qualifier doesn't make sense.

Well, I guess a `Instance#activeHandles()` method would be one step closer to what I suggested, but it feels wonky.

Reply:
> I suppose you're saying that because the bean would (per spec) lose the `@Default` qualifier?

Yeah, that's what I meant.

It is true that we could special-case the `@Active` qualifier (like `@Any` and `@Named` are, and to some extent our `@All`), but I'd like to avoid that just because special cases in this are are super surprising (cf. `@Named`).

It does have certain appeal, though. I'll think some more about it.

Reply:
FWIW, everything works without that `@Active` "pseudo-qualifier" in my PR, and it works fine in extension code. But I must admit it feels a bit complicated when we suggest it to application developers.

See for example [here](https://quarkus-pr-main-41929-preview.surge.sh/version/main/guides/datasource#datasource-active), in the CDI producer definition:

```java
public class MyProducer {
    @Inject
    @DataSource("pg")
    InjectableInstance<AgroalDataSource> pgDataSourceBean; 

    @Inject
    @DataSource("oracle")
    InjectableInstance<AgroalDataSource> oracleDataSourceBean;

    @Produces 
    @ApplicationScoped
    public AgroalDataSource dataSource() {
        if (pgDataSourceBean.getHandle().getBean().isActive().result()) { 
            return pgDataSourceBean.get();
        } else if (oracleDataSourceBean.getHandle().getBean().isActive().result()) { 
            return oracleDataSourceBean.get();
        } else {
            throw new RuntimeException("No active datasource!");
        }
    }
}
```

I suppose we could say it's "not great, not terrible" :)

But with an `@Active` qualifier, and if that qualifier could be made to work in static injection points as well... we could get simpler code:

```java
public class MyProducer {
    @Inject
    @DataSource("pg")
    @Active
    Instance<AgroalDataSource> pgDataSourceBean; 

    @Inject
    @DataSource("oracle")
    @Active
    Instance<AgroalDataSource> oracleDataSourceBean;

    @Produces 
    @ApplicationScoped
    public AgroalDataSource dataSource() {
        if (pgDataSourceBean.isResolvable()) { 
            return pgDataSourceBean.get();
        } else if (oracleDataSourceBean.isResolvable()) { 
            return oracleDataSourceBean.get();
        } else {
            throw new RuntimeException("No active datasource!");
        }
    }
}
```

Or even just (though I'm not entirely sure about this one -- I suspect static injection combined with `@Active` may not be straightforward to implement):

```java
public class MyProducer {
    @Inject
    @Active
    AgroalDataSource activeDataSource;

    @Produces 
    @ApplicationScoped
    public AgroalDataSource dataSource() {
        return activeDataSource;
    }
}
```

Reply:
> I suppose we could say it's "not great, not terrible" :)

Agreed :-)

The `@Active` qualifier is appealing, but I'd rather leave it out of the initial iteration. We should be able to add it later.